scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -279,9 +279,121 @@
+                  (recur instance-id->progress-state'))
+                (do
+                  (meters/mark! progress-aggregator-drop-rate)
+                  (counters/inc! progress-aggregator-drop-count)
+                  (log/debug "Dropping" data "as threshold has been reached")
+                  (recur instance-id->progress-state))))
+
+            query-state-chan
+            (let [{:keys [mode response-chan]} data]
+              (condp = mode
+                :query
+                (do
+                  (async/>! response-chan instance-id->progress-state)
+                  (recur instance-id->progress-state))
+
+                :exit
+                (do
+                  (when response-chan
+                    (async/>! response-chan :exited))
+                  (log/info "Exiting progress update aggregator"))))))))
+    {:progress-aggregator-chan progress-aggregator-chan
+     :query-state-chan query-state-chan}))
+
+(histograms/defhistogram [cook-mesos scheduler progress-updater-pending-states])
+(meters/defmeter [cook-mesos scheduler progress-updater-publish-rate])
+(timers/deftimer [cook-mesos scheduler progress-updater-publish-duration])
+(meters/defmeter [cook-mesos scheduler progress-updater-tx-rate])
+(timers/deftimer [cook-mesos scheduler progress-updater-tx-duration])
+
+(defn progress-update-transactor
+  "Launches a long running go block that triggers transacting the latest aggregated instance-id->progress-state to
+   datomic whenever there is a message on the progress-updater-trigger-chan.
+   No more than batch-size facts are updated in individual datomic transactions.
+   It returns a progress-state-chan which can be used to send messages about the latest instance-id->progress-state.
+   If no such message is on the progress-state-chan, then no datomic interactions occur."
+  [progress-updater-trigger-chan batch-size conn]
+  (log/info "Starting progress update transactor")
+  (let [progress-state-chan (async/chan (async/sliding-buffer 1))]
+    (async/go
+      (loop []
+        (let [{:keys [response-chan]} (async/<! progress-updater-trigger-chan)
+              instance-id->progress-state (async/<! progress-state-chan)]
+          ;; TODO metrics for map size, datomic queries, etc.
+          (log/info "Received" (count instance-id->progress-state) "in progress-update-transactor")
+          (if instance-id->progress-state
+            (do
+              (histograms/update! progress-updater-pending-states (count instance-id->progress-state))
+              (meters/mark! progress-updater-publish-rate)
+              (timers/time!
+                progress-updater-publish-duration
+                (doseq [instance-id->progress-state-partition
+                        (partition-all batch-size instance-id->progress-state)]
+                  (try
+                    (let [transactions
+                          (reduce (fn progress-update-transactor-reducer
```

Comment:
I'd prefer using `mapcat` here. When I see a `reduce` I need to read it carefully because `reduce` can do anything. Using `mapcat` better signals the intent